### PR TITLE
ENH Make JS Errors work like other special JS types

### PR DESCRIPTION
**Pull request by hoodmane**
### Description

Up to this point, we've used this dynamic subclassing method for producing JsProxies for everything but errors. For errors, we make a wrapper which is not a JsProxy that inherits from Exception and give the wrapper a `js_error` attribute that points to an actual error. We also make python2js know about this wrapper so it can unwrap it. But the raw js_error object is a bit weird. There isn't anything terrible about this situation but it is mildly unsatisfying.

This changes it so that errors subclass both `JsProxy` and `Exception`. To do this we need to:
1. ensure that they have compatible memory layouts, and
2. convince Python that they have compatible memory layouts

I switched to using a union for the different subtypes of `JsProxy` that need extra space: `JsCallable`, `JsBuffer`, and `JsError`. We need js to be at the end so it won't get in the way of the `BaseException` memory layout. I added _Static_asserts to double check that the memory layouts do in fact agree.

To convince Python that they have compatible memory layouts, we have to temporarily tell it that `JsProxy` is a subtype of `BaseException.` To do this, we just set `JsProxy.__mro__ = (BaseException,)` before creating the type and then restore it afterward.

### Checklists

- [x] Add a [CHANGELOG](https://github.com/pyodide/pyodide/blob/main/docs/project/changelog.md) entry

### CODE REVIEW

**Comment by hoodmane** [comment]
@rth @ryanking13 when one of you gets a chance if you could review this I would appreciate it.

**Comment by rth** [comment]
Thanks for working on this! Will try to review it in the next few days, unless Gyeongjae gets to it before )

**Review by ryanking13 — APPROVED** [review]
Thanks for working on this @hoodmane! It generally looks good to me.

File: src/tests/test_pyodide.py [hunk 1]
```diff
@@ -1337,7 +1337,7 @@ def test_pickle_jsexception(selenium):
 
 
 def test_raises_jsexception(selenium):
-    from pytest_pyodide.pyodide import JsException
+    from pyodide.ffi import JsException
```
[the remark this answers is not in the document]
Comment (hoodmane) [reply src/tests/test_pyodide.py:1370]: Perhaps we should update `pytest_pyodide` to be aware of this...

**Comment by hoodmane** [comment]
Thanks for the review @ryanking13!

File: src/tests/test_pyodide.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ryanking13) [reply src/tests/test_pyodide.py:1370]: Yeah, but I guess we should still need to keep the JsException class in `pytest_pyodide` for backward compatibility...

**Review by ryanking13 — APPROVED** [review]

**Pull request merged**